Source/FulcrumFS.Videos/VideoProcessor.cs

```diff
@@ -0,0 +1,236 @@
+            }
+
+            IReadOnlyList<AudioCodec> result = [.. value];
+
+            if (result.Count is 0)
+                throw new ArgumentException("Codecs cannot be empty.", nameof(value));
+
+            if (result.Any((x) => x is null))
+                throw new ArgumentException("Codecs cannot contain null values.", nameof(value));
+
+            if (result.Distinct().Count() != result.Count)
+                throw new ArgumentException("Codecs cannot contain duplicates.", nameof(value));
+
+            field = result;
+        }
+    }
+
+    /// <summary>
+    /// Gets the source media container format for this mapping (it matches any of these).
+    /// If <see langword="null" />, matches any supported container format.
+    /// Default is <see langword="null" />.
+    /// </summary>
+    public IReadOnlyList<MediaContainerFormat>? SourceMediaContainerFormats
+    {
+        get;
+        init
+        {
+            if (value is null)
+            {
+                field = null;
+                return;
+            }
+
+            IReadOnlyList<MediaContainerFormat> result = [.. value];
+
+            if (result.Count is 0)
+                throw new ArgumentException("Formats cannot be empty.", nameof(value));
+
+            if (result.Any((x) => x is null))
+                throw new ArgumentException("Formats cannot contain null values.", nameof(value));
+
+            if (result.Distinct().Count() != result.Count)
+                throw new ArgumentException("Formats cannot contain duplicates.", nameof(value));
+
+            field = result;
+        }
+    }
+
+    /// <summary>
+    /// Gets the video stream processing options for the video streams in the source video.
+    /// Default is a new instance of <see cref="VideoStreamProcessingOptions" /> that specifies always re-encoding to a standardised H.264 stream.
+    /// </summary>
+    public VideoStreamProcessingOptions VideoStreamOptions { get; init; } = new VideoStreamProcessingOptions();
+
+    /// <summary>
+    /// Gets the audio stream processing options for the audio streams in the source video.
+    /// Default is a new instance of <see cref="AudioStreamProcessingOptions" /> that specifies always re-encoding to a standardised AAC stream.
+    /// </summary>
+    public AudioStreamProcessingOptions AudioStreamOptions { get; init; } = new AudioStreamProcessingOptions();
+
+    /// <summary>
+    /// Gets the result media container format for this mapping, or null to use the same as the input.
+    /// If <see langword="null" />, the container format of the source video is preserved (when possible, i.e., if re-encoding, or otherwise requesting to
+    /// modify things like metadata, it will automatically default to another container if unsupported for writing).
+    /// Default is a list containing <see cref="MediaContainerFormat.MP4" />.
+    /// </summary>
+    public IReadOnlyList<MediaContainerFormat>? ResultMediaContainerFormat

```

Comment:
Same as above, I don't think we should allow null here - init to a `MediaContainerFormat.AllResultFormats` list that contains all supported output formats (with MP4 as the first one in the list so it is "preferred" if conversion is necessary).

Reply:
Or we could have default instances that provide both behaviors - `MP4H264Instance` is always consistent, `DefaultMaxCompatibilityInstance` can be updated between major releases to be a current safe default for all reasonable modern devices, etc.

We could even make the processor ctor internal and force selection from one of the static default starting points, in which case we would have one that's `PreserveInputInstance`. I kind of like that idea. I may do something similar with images.

Reply:
You won't be able to make it a `record`, unless you also make the base type/s a record. I also think it is a bit of an abuse to make the processor types `record`s (and it also locks you into a design where you won't want to cache anything on the instances, as that would be copied if you ever did `with`, and also it would be printed out whenever you do ToString, checked for equality, etc. - not certain if you can easily make it not do all of those), but it could make sense to make all of the config stuff `record`s if we seperated it back out & used some EquatableImmutableArray-like thing - I don't mind that idea in combination with your private ctor + default provided configs idea tbh.